src/rust/ide/lib/span-tree/src/node/kind.rs

```diff
@@ -0,0 +1,303 @@
+//! SpanTree node kind implementation. Each node can be of a different kind (shape) and can contain
+//! different information.
+
+use crate::prelude::*;
+
+use crate::ArgumentInfo;
+
+
+
+// ============
+// === Kind ===
+// ============
+
+/// An enum describing kind of node.
+#[derive(Clone,Debug,Eq,PartialEq)]
+pub enum Kind {
+    /// A root of the expression tree.
+    Root,
+    /// A node chained with parent node. See crate's docs for more info about chaining.
+    Chained,
+    /// A node representing operation (operator or function) of parent Infix, Section or Prefix.
+    Operation,
+    /// A node being a target (or "self") parameter of parent Infix, Section or Prefix.
+    This(This),
+    /// A node being a normal (not target) parameter of parent Infix, Section or Prefix.
+    Argument(Argument),
+    /// A node being a placeholder for inserting new child to Prefix or Operator chain. It should
+    /// not have children, but can be assigned with a span representing the number of spaces between
+    /// AST tokens. For example, given expression `foo   bar`, the span assigned to the
+    /// `InsertionPoint` between `foo` and `bar` should be set to 3.
+    InsertionPoint(InsertionPoint),
+}
+
+
+// === Smart Kind Constructors ===
+
+#[allow(missing_docs)]
+impl Kind {
+    pub fn this            () -> This           { default() }
+    pub fn argument        () -> Argument       { default() }
+    pub fn insertion_point () -> InsertionPoint { default() }
+}
```

Comment:
Is `Kind::this()` so better than `kind::This::default()`?

Reply:
As agreed via phone, yes it is in certain situations :)